File: frontend/app/models/category.js
import Model, { attr, belongsTo, hasMany } from '@ember-data/model';

// KEEP UPDATED WITH:
//
//   app/serializers/category_serializer.rb
//
export default class CategoryModel extends Model {
  @attr name;
  @attr slug;
  @attr description;
  @hasMany('category', { inverse: 'parent' }) children;
  @belongsTo('category', { inverse: 'children' }) parent;

  // Relationships
  @hasMany('have-post') havePosts;
  @hasMany('need-post') needPosts;
}

File: app/models/category.rb
class Category < ApplicationRecord
  belongs_to :parent, class_name: "Category", optional: true
  has_many :children, class_name: "Category", foreign_key: "parent_id"

  # Relationships
  has_many :need_posts
  has_many :have_posts

  # Validations
  validates :slug, uniqueness: true

  # Callbacks
  def before_create
    if @slug.nil?
      @slug = self.parent_slug ? "#{self.parent_slug}.#{@name.dasherize}" : @name.dasherize
    end
  end

end
